src/engine/PathSearch.cpp

```diff
@@ -0,0 +1,391 @@
+    info.addDetail("Time to search paths", searchTime.count());
+    info.addDetail("Time to fill result table", fillTime.count());
+  }
+
+  return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
+};
+
+// _____________________________________________________________________________
+VariableToColumnMap PathSearch::computeVariableToColumnMap() const {
+  return variableColumns_;
+};
+
+std::span<const Id> PathSearch::handleSearchSide(
+    const SearchSide& side, const std::optional<TreeAndCol>& binding) const {
+  std::span<const Id> ids;
+  bool isVariable = std::holds_alternative<Variable>(side);
+  if (isVariable && binding.has_value()) {
+    ids = binding->first->getResult()->idTable().getColumn(binding->second);
+  } else if (isVariable) {
+    return {};
+  } else {
+    ids = std::get<std::vector<Id>>(side);
+  }
+  return ids;
+}
+
+// _____________________________________________________________________________
+std::vector<Path> PathSearch::findPaths(
+    const Id source, const std::unordered_set<uint64_t>& targets,
+    const BinSearchWrapper& binSearch) const {
+  std::forward_list<Edge> edgeStack;
+  Path currentPath;
+  std::unordered_map<uint64_t, std::vector<Path>> pathCache;
+  std::unordered_set<uint64_t> visited;
+
+  auto addToCache =
+      [](std::unordered_map<uint64_t, std::vector<Path>>& pathCache,
+         const Path& path, size_t stopIndex) {
+        for (size_t i = 0; i < stopIndex; i++) {
+          const auto& edge = path.edges_[i];
+          auto startIndex = edge.start_.getBits();
+          pathCache.try_emplace(startIndex, std::vector<Path>());
+          pathCache[startIndex].push_back(path.startingAt(i));
+        }
+      };
+
+  visited.insert(source.getBits());
+  for (auto edge : binSearch.outgoingEdes(source)) {
+    edgeStack.push_front(std::move(edge));
+  }
+
+  while (!edgeStack.empty()) {
+    auto edge = edgeStack.front();
+    edgeStack.pop_front();
+
+    while (!currentPath.empty() && edge.start_ != currentPath.end()) {
+      currentPath.pop_back();
+    }
+
+    currentPath.push_back(edge);
+
+    // TODO clean this up
+    if (pathCache.contains(edge.end_.getBits())) {
+      for (auto subPath : pathCache[edge.end_.getBits()]) {
+        if (subPath.first() == currentPath.first()) {
+          addToCache(pathCache, currentPath, currentPath.size());
+        } else {
+          auto fullPath = currentPath.concat(subPath);
+          addToCache(pathCache, fullPath, currentPath.size());
+        }
+      }
+      continue;
+    }
+
+    if (targets.empty() || targets.contains(edge.end_.getBits())) {
+      addToCache(pathCache, currentPath, currentPath.size());
+    }
+
+    if (!visited.contains(edge.end_.getBits())) {
+      for (auto outgoingEdge : binSearch.outgoingEdes(edge.end_)) {
+        edgeStack.push_front(outgoingEdge);
+      }
+    }
+
+    visited.insert(edge.end_.getBits());
+  }
+
+  return pathCache[source.getBits()];
```

Comment:
I think for this function at all we should aim for the most efficient implementation thinkable.
And that is, I think, one that directly stores to the IdTable and has very simple path creations. We can talk about this again if you want.

One thing we have to think about: In the case where you have pairs of [source, target], the path cache has to be invalidated the set of targets changes. So you basically should

Reply:
Okay, 
 I just thought of an idea:
 1. You store the complete subtree in a deterministic and sorted way.
 2. Then the `Edge` type can just be a single rowIndex into that table (or the individual spans for start and end nodes, which is the same). That way you don't really have the problem with the overhead, and the edge Properties etc, as you only have to touch all this when you actually materialize the result. This should all be really efficient (no allocations etc).